spanner/transaction.go

```diff
@@ -1072,20 +1096,72 @@ func (t *ReadWriteTransaction) batchUpdateWithOptions(ctx context.Context, stmts
 
 // acquire implements txReadEnv.acquire.
 func (t *ReadWriteTransaction) acquire(ctx context.Context) (*sessionHandle, *sppb.TransactionSelector, error) {
-	ts := &sppb.TransactionSelector{
-		Selector: &sppb.TransactionSelector_Id{
-			Id: t.tx,
-		},
+	for {
+		t.mu.Lock()
+		switch t.state {
+		case txClosed:
+			t.mu.Unlock()
+			return nil, nil, errTxClosed()
+		case txNew:
+			// State transit to txInit so that only one TransactionSelector::begin
+			// is accepted.
+			t.state = txInit
+			sh := t.sh
+			ts := &sppb.TransactionSelector{
+				Selector: &sppb.TransactionSelector_Begin{
+					Begin: &sppb.TransactionOptions{
+						Mode: &sppb.TransactionOptions_ReadWrite_{
+							ReadWrite: &sppb.TransactionOptions_ReadWrite{},
+						},
+					},
+				},
+			}
+			t.mu.Unlock()
+			return sh, ts, nil
+		case txInit:
+			if t.tx == nil {
+				// Wait for a transaction ID to become ready.
+				txReadyOrClosed := t.txReadyOrClosed
+				t.mu.Unlock()
+				select {
+				case <-txReadyOrClosed:
+					// Need to check transaction state again.
+					continue
+				case <-ctx.Done():
+					// The waiting for initialization is timeout, return error
+					// directly.
+					return nil, nil, errTxInitTimeout()
+				}
+			}
+			continue
+		case txActive:
+			sh := t.sh
+			ts := &sppb.TransactionSelector{
+				Selector: &sppb.TransactionSelector_Id{
+					Id: t.tx,
+				},
+			}
+			t.mu.Unlock()
+			return sh, ts, nil
+		}
+		state := t.state
+		t.mu.Unlock()
+		return nil, nil, errUnexpectedTxState(state)
+	}
+}
+
+func (t *ReadWriteTransaction) setTransactionID(tx transactionID) {
+	if tx == nil {
+		return
 	}
 	t.mu.Lock()
 	defer t.mu.Unlock()
-	switch t.state {
-	case txClosed:
-		return nil, nil, errTxClosed()
-	case txActive:
-		return t.sh, ts, nil
+	if t.state == txInit {
```

Comment:
When is this function called with `t.state != txInit`? And is it then reasonable that it is effectively a no-op?